### ISSUE TITLE
Add use case for error stacks

### ISSUE THREAD

**Issue by spion** [nodejs/promise-use-cases#8]
* user using testing library such as e.g. Mocha in node
* an error is generated in a promise deep within several async calls
* stack trace printed by mocha only contains stack after last await

another use case is logging promise errors in production, that suffers from the same issue.

Not sure if we can use the Inspector API in this scenario: https://nodejs.org/api/inspector.html that seems super awkward.

Would be much better if we could get the hooks to stitch a proper async stack trace, or even easier `Promise.getLongStackTrace(rejectedPromise)` similar to `Error.captureStackTrace`

Another API that worked nicely with generators for a short while was https://github.com/bjouhier/galaxy-stack which let you get the individual stack frame of currently paused generators, it would be nice to get something like that for both generators and/or async functions.

**Comment by benjamingr** [comment]
This is an excellent suggestion - can you make a PR?

**Comment by spion** [comment]
I'll try, and also pinging @bjouhier whose code was way ahead of its time and node wasn't ready yet :grinning:. @bjouhier if you have the time any input you have would be greatly appreciated, lets try and make async/await properly usable.

**Comment by benjamingr** [comment]
Yes please! Ping anyone else you think has insights into this as well :)

**Comment by bjouhier** [comment]
Thanks @spion. `streamline.js` has always supported full async stacktraces, with little overhead (compared to libraries that extract the missing frames from an `Error` object).

This was possible because streamline is a preprocessor. The `galaxy` library also supported full stacktraces although it did not require any preprocessing but I had to write a binary module to add the missing API to generators (and this API was an ugly hack with V8 internals so I had to drop it).

Full stacktraces can also be implemented around async/await but I don't see how this can be done with a pure JS library. The reason is that the extra code needs to be associated to the `await` keywords, not to all promises. And userland does not see the `await` keywords, unless it transpiles.

I have quickly hacked [a small gist](https://gist.github.com/bjouhier/9882719d30fb171cdabdfdf55c05ac4f) to demo how this could be done with a transpiler pass but the proper way to implement it should be via the JS compiler and the internals of the `Promise` / `Error` classes. I have given some hints in the gist comment.

*Side note*: I don't really need this feature any more so my gist and comments are not backed by a real implementation. Take them with a pinch of salt. I am now writing all my node.js code with [f-promise](https://github.com/Sage/f-promise), a small library that gives me the benefits of async/await  without the keyword pollution, and that works seamlessly with all JS APIs (array.map, array.reduce, string.replace, promises, callbacks, etc.). And, of course, it supports full async stacktraces (with zero overhead) and CLS.

I just wish that the single-frame-continuation constraint be lifted some day and that we'd get proper support for real (stackful) coroutines in JavaScript. My take after 7+ years of async/await coding (streamline.js was just an early implementation of async/await, with a different syntax).
